### PR TITLE
Optmize renderState and its related functions

### PR DESCRIPTION
**Pull request by laisspportugal**
**Description**

This pull request aims to optimize `renderState` and its related functions to enhance readability and maintainability.  

### **Key Changes**:  
- **Generalization**:  
  Consolidated repetitive code to improve efficiency and maintainability.  

- **Function Decomposition**:  
  Split large functions into smaller, more focused functions for better readability and reuse.  

- **Unit Test Coverage**:  
  - Added additional unit tests to improve test coverage.  
  - Refactored some existing unit tests using builders

**Checklist**

- [x] The web version was tested and it is running ok
- [x] The desktop version was tested and it is running ok
- [x] This change is covered by unit tests
- [x] Files constants.ts, types.ts and *.style.ts have been checked and relevant code snippets have been relocated

### CODE REVIEW

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      lastSeekTime: BasicBuilder.number(),
+      currentTime: RosTimeBuilder.time(),
+      endTime: RosTimeBuilder.time(),
+      startTime: RosTimeBuilder.time(),
+      isPlaying: true,
+      messages: [],
+      speed: 1,
+      topics: [],
+      topicStats: new Map(),
+      totalBytesReceived: BasicBuilder.number(),
```
Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:37]: Order parameters a-z
tip: Select lines in VSCode > Ctrl+P > type: ">Sort lines"

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    });
+  }
+
+  public static datatypes(props: Partial<RosDatatypes> = {}): RosDatatypes {
+    return defaults<RosDatatypes>(
+      props,
+      BasicBuilder.genericMap(PlayerStateBuilder.optionalMessageDefinition),
+    );
+  }
+}
```
Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:56]: Add these builders to be reusable above. 

```suggestion

  public static topicStats(props: Partial<TopicStats> = {}): TopicStats {
    return defaults<TopicStats>(props, {
      firstMessageTime: RosTimeBuilder.time(),
      lastMessageTime: RosTimeBuilder.time(),
      numMessages: BasicBuilder.number(),
    });
  }

  public static topic(props: Partial<Topic> = {}): Topic {
    return defaults<Topic>(props, {
      aliasedFromName: BasicBuilder.string(),
      name: BasicBuilder.string(),
      schemaName: BasicBuilder.string(),
    });
  }

  public static topics(count = 3): Topic[] {
    return BasicBuilder.multiple(PlayerStateBuilder.topic, count);
  }
```

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  public static activeData(props: Partial<PlayerStateActiveData> = {}): PlayerStateActiveData {
+    return defaults<PlayerStateActiveData>(props, {
+      datatypes: PlayerStateBuilder.datatypes(),
+      lastSeekTime: BasicBuilder.number(),
+      currentTime: RosTimeBuilder.time(),
+      endTime: RosTimeBuilder.time(),
+      startTime: RosTimeBuilder.time(),
+      isPlaying: true,
+      messages: [],
+      speed: 1,
```
Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:34]: Already exist messageEvents builder

```suggestion
      messages: MessageEventBuilder.messageEvents(),
```

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts (hunk 1 above)
Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:37]: Reuse topics implemented as review suggestion.

```suggestion
      topics: PlayerStateBuilder.topics(),
```

Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:37]: Reuse topicStats implemented as review suggestion.

```suggestion
      topicStats: BasicBuilder.genericMap(PlayerStateBuilder.topicStats),
```

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+class PlayerStateBuilder {
+  public static playerState(props: Partial<PlayerState> = {}): PlayerState {
+    return defaults<PlayerState>(props, {
+      presence: BasicBuilder.sample(PlayerPresence),
+      progress: {},
+      capabilities: [],
+      profile: BasicBuilder.string(),
+      playerId: BasicBuilder.string(),
+      activeData: PlayerStateBuilder.activeData(),
+    });
```
Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:23]: Please order

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  PlayerPresence,
+  PlayerState,
+  PlayerStateActiveData,
+} from "@lichtblick/suite-base/players/types";
+import BasicBuilder from "@lichtblick/suite-base/testing/builders/BasicBuilder";
+import RosTimeBuilder from "@lichtblick/suite-base/testing/builders/RosTimeBuilder";
+import { defaults } from "@lichtblick/suite-base/testing/builders/utilities";
+import { OptionalMessageDefinition, RosDatatypes } from "@lichtblick/suite-base/types/RosDatatypes";
+
+class PlayerStateBuilder {
```
Comment (luluiz) [inline packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:14]: @laisspportugal I've just noticed that **PlayerBuilder** already implements the **PlayerState** data.
So, I'd say to remove this file e reuse from **PlayerBuilder**

Comment (luluiz) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:14]: ... and my last comments in this PR became irrelevant

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts (hunk 4 above)
Comment (luluiz) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:23]: ignore, please read my last [comment](https://github.com/Lichtblick-Suite/lichtblick/pull/266#discussion_r1850184006)

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts (hunk 1 above)
Comment (luluiz) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:37]: ignore, please read my last [comment](https://github.com/Lichtblick-Suite/lichtblick/pull/266#discussion_r1850184006)

Comment (luluiz) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:37]: ignore, please read my last [comment](https://github.com/Lichtblick-Suite/lichtblick/pull/266#discussion_r1850184006)

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts (hunk 3 above)
Comment (luluiz) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:34]: ignore, please read my last [comment](https://github.com/Lichtblick-Suite/lichtblick/pull/266#discussion_r1850184006)

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts (hunk 1 above)
Comment (luluiz) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts:37]: ignore, please read my last [comment](https://github.com/Lichtblick-Suite/lichtblick/pull/266#discussion_r1850184006)

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts [hunk 6]
```diff
@@ -108,32 +120,40 @@ describe("renderState", () => {
...
+    });
+
+    expect(state).toEqual({
+      topics: [{ name: "test", schemaName: "schema", datatype: "schema" }],
+    });
+  });
+
   it("should provide stable time values", () => {
     const buildRenderState = initRenderStateBuilder();
+    const playerState = PlayerStateBuilder.playerState();
```
Comment (luluiz) [inline packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts:146]: Reuse from `PlayerBuilder`

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts [hunk 7]
```diff
@@ -258,86 +278,34 @@ describe("renderState", () => {
...
-                      schemaName: "schema",
-                      receiveTime: { sec: 8, nsec: 0 },
-                      sizeInBytes: 1,
-                      message: { from: "allFrames" },
-                    },
-                  ],
-                },
-              },
-            ],
+    const playerState = PlayerStateBuilder.playerState();
```
Comment (luluiz) [inline packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts:281]: Reuse from `PlayerBuilder`

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts [hunk 8]
```diff
@@ -1125,4 +1045,131 @@ describe("renderState", () => {
...
+  });
+
+  it("should renderStateField when activeData contains parameters", async () => {
+    const buildRenderState = initRenderStateBuilder();
+    const stableConversionInputs = {
+      sortedTopics: [],
+      subscriptions: [{ topic: "test" }],
+      messageConverters: [],
+    };
+    const activeData = PlayerStateBuilder.activeData({
```
Comment (luluiz) [inline packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts:1081]: Reuse from `PlayerBuilder`

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts [hunk 9]
```diff
@@ -1125,4 +1045,131 @@ describe("renderState", () => {
...
+      globalVariables: {},
+      hoverValue: undefined,
+      sharedPanelState: {},
+      ...stableConversionInputs,
+    });
+
+    expect(state).toEqual({
+      parameters: new Map(),
+    });
+  });
```
Comment (luluiz) [inline packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts:1072]: Example using the `setup()`

```suggestion
  it("should update renderStateField when watchedFields contains parameters", async () => {
    const { buildRenderState, input } = setup();
    _.merge(input, {
      messageConverters: [],
      sortedTopics: [],
      subscriptions: [BasicBuilder.genericMap],
      watchedFields: new Set(["parameters"]),
    });

    const state = buildRenderState(input);

    expect(state).toEqual({
      parameters: new Map(),
    });
  });
```

Comment (luluiz) [inline packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts:1072]: Example using the `setup()`

```suggestion
  it("should update renderStateField when watchedFields contains parameters", async () => {
    const { buildRenderState, input } = setup();
    _.merge(input, {
      messageConverters: [],
      sortedTopics: [],
      subscriptions: [BasicBuilder.genericMap],
      watchedFields: new Set(["parameters"]),
    });

    const state = buildRenderState(input);

    expect(state).toEqual({
      parameters: new Map(),
    });
  });
```

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts [hunk 10]
```diff
@@ -1125,4 +1045,131 @@ describe("renderState", () => {
...
+      globalVariables: {},
+      hoverValue: undefined,
+      sharedPanelState: {},
+      ...stableConversionInputs,
+    });
+
+    expect(state).toEqual({
+      parameters: activeData.parameters,
+    });
+  });
```
Comment (luluiz) [inline packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts:1100]: Second example using `setup()`
```suggestion
  it("should renderStateField when activeData contains parameters", async () => {
    const { buildRenderState, input } = setup();
    const activeData = PlayerStateBuilder.activeData({
      parameters: BasicBuilder.genericMap<string>(BasicBuilder.string),
    });
    const playerState = PlayerStateBuilder.playerState({ activeData });
    _.merge(input, {
      messageConverters: [],
      playerState,
      sortedTopics: [],
      subscriptions: [BasicBuilder.genericMap],
      watchedFields: new Set(["parameters"]),
    });
    const state = buildRenderState(input);

    expect(state).toEqual({
      parameters: activeData.parameters,
    });
  });
  ```

**Review by luluiz — CHANGES_REQUESTED** [review]

File: packages/suite-base/src/testing/builders/PlayerStateBuilder.ts (hunk 5 above)
Comment (laisspportugal) [reply packages/suite-base/src/testing/builders/PlayerStateBuilder.ts]: Fixed

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts (hunk 6 above)
Comment (laisspportugal) [reply packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts]: Fixed

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts (hunk 7 above)
Comment (laisspportugal) [reply packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts]: Fixed

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts (hunk 8 above)
Comment (laisspportugal) [reply packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts]: Fixed

File: packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts [hunk 11]
```diff
@@ -1125,4 +1045,131 @@ describe("renderState", () => {
...
+    const buildRenderState = initRenderStateBuilder();
+    const stableConversionInputs = {
+      sortedTopics: [],
+      subscriptions: [{ topic: "test" }],
+      messageConverters: [],
+    };
+    const activeData = PlayerStateBuilder.activeData({
+      parameters: BasicBuilder.genericMap<string>(BasicBuilder.string),
+    });
+    const playerState = PlayerStateBuilder.playerState({ activeData });
```
[the remark this answers is not in the document]
Comment (laisspportugal) [reply packages/suite-base/src/components/PanelExtensionAdapter/renderState.test.ts]: Fixed

**Review by luluiz — APPROVED** [review]

**Pull request merged**